simulate/sam.go

```diff
@@ -0,0 +1,143 @@
+package simulate
+
+import (
+	"fmt"
+	"github.com/vertgenlab/gonomics/cigar"
+	"github.com/vertgenlab/gonomics/dna"
+	"github.com/vertgenlab/gonomics/numbers"
+	"github.com/vertgenlab/gonomics/sam"
+	"math/rand"
+	"strings"
+)
+
+// IlluminaSam generates pseudorandom paired reads evenly distributed along the input ref sequence using typical illumina parameters.
+func IlluminaSam(refName string, ref []dna.Base, numPairs int) []sam.Sam {
+	return simulatePairedSam(refName, ref, numPairs, 150, 50, 50)
+}
+
+// simulatePairedSam generates a pair of sam reads randomly distributed across the input ref.
+func simulatePairedSam(refName string, ref []dna.Base, numPairs, readLen, avgInsertSize int, avgInsertSizeStdDev float64) []sam.Sam {
+	reads := make([]sam.Sam, numPairs*2)
+
+	var insertSize, midpoint, startFor, startRev, endFor, endRev int
+	for i := 0; i < len(reads); i += 2 {
+		insertSize = int(numbers.SampleInverseNormal(float64(avgInsertSize), avgInsertSizeStdDev))
+		midpoint = numbers.RandIntInRange(-2*readLen, len(ref)+(2*readLen))
+		startFor = midpoint - (readLen + (insertSize / 2))
+		endFor = startFor + readLen
+		startRev = midpoint + (insertSize / 2)
+		endRev = startRev + readLen
+
+		reads[i] = generateSamReadNoFlag(fmt.Sprintf("%s_Read:%d", refName, i/2), refName, ref, startFor, endFor, true)
+		reads[i+1] = generateSamReadNoFlag(fmt.Sprintf("%s_Read:%d", refName, i/2), refName, ref, startRev, endRev, false)
+		if reads[i].Cigar == nil && reads[i+1].Cigar == nil {
+			i -= 2 // retry
+			continue
+		}
+		addPairedFlags(&reads[i], &reads[i+1])
+		if reads[i].Cigar != nil && reads[i+1].Cigar != nil {
+			reads[i].RNext = "="
+			reads[i+1].RNext = "="
+		} else {
+			reads[i].RNext = reads[i+1].RName
+			reads[i+1].RNext = reads[i].RName
+		}
+
+		reads[i].PNext = reads[i+1].Pos
+		reads[i+1].PNext = reads[i].Pos
+	}
+	return reads
+}
+
+// generateSamReadNoFlag generates a sam record for the input position.
+// Soft clips sequence that is off template and does not generate Flag, RNext, or PNext.
+func generateSamReadNoFlag(readName string, refName string, ref []dna.Base, start, end int, isFor bool) sam.Sam {
+	var s sam.Sam
+	s.QName = readName
+	s.Seq = make([]dna.Base, end-start)
+	// generate qual
+	var bldr strings.Builder
+	for range s.Seq {
+		bldr.WriteRune(rune(numbers.RandIntInRange(30, 40) + 33)) // high quality seq + ascii offset
+	}
+	s.Qual = bldr.String()
+
+	// check if unmapped
+	if end < 0 || start > len(ref) {
+		s.RName = "*"
+		for i := range s.Seq {
+			s.Seq[i] = dna.Base(numbers.RandIntInRange(0, 4))
+		}
+		return s
+	}
+
+	s.MapQ = uint8(numbers.RandIntInRange(30, 40))
+	s.RName = refName
+
+	// generate random seq if off template
+	var realSeqStartIdx, realSeqEndIdx int
+	for realSeqStartIdx = start; realSeqStartIdx < 0; realSeqStartIdx++ {
+		s.Seq[realSeqStartIdx-start] = dna.Base(numbers.RandIntInRange(0, 4))
+	}
+	for realSeqEndIdx = end; realSeqEndIdx > len(ref); realSeqEndIdx-- {
```

Comment:
Should this be ">=" instead of ">"?

Reply:
The interval is `[realSeqStartIdx, realSeqEndIdx)` so `realSeqEndIdx == len(ref)` is a valid index so I believe > is correct.